### PR TITLE
Several fixes to support long-lived persistent connections

### PR DESCRIPTION
**Pull request by tgraf**

### CODE REVIEW

**Comment by tgraf** [comment]
test-me-please

**Review by rlenglet — APPROVED** [review]

**Review by manalibhutiyani — APPROVED** [review]
Approving for Kafka bits, they look fine.

**Comment by rlenglet** [comment]
After https://github.com/cilium/cilium/pull/2781 is in, we'll be able to set the socket options similarly for Envoy, in `envoy/cilium_bpf_metadata.cc`.

**Comment by tgraf** [comment]
test-me-please

**Comment by tgraf** [comment]
test-me-please

File: pkg/proxy/dialer.go [hunk 1]
```diff
@@ -66,6 +81,11 @@ func ciliumDialer(identity int, network, address string) (net.Conn, error) {
 		return nil, fmt.Errorf("unable to connect: %s", err)
 	}
 
+	if err := setKeepAlive(c); err != nil {
```
Comment (rlenglet) [inline pkg/proxy/dialer.go]: We also need to set those options on request connections, in `pkg/proxy/socket.go`'s `proxySocket.Accept`.

Comment (rlenglet) [inline pkg/proxy/dialer.go]: Shouldn't we also call `TCPConn.SetLinger` with a delay consistent with `proxyConnectionCloseTimeout`, in order to bound the time it takes to close a connection?
Again, we'd need to do that on request connections as well.

File: bpf/lib/common.h [hunk 2]
```diff
@@ -398,7 +398,10 @@ struct ct_state {
 	__u32 src_sec_id;
 };
 
-#define PROXY_DEFAULT_LIFETIME	360
+/* Lifetime of a proxy redirection entry is an entire day. All proxies should
+ * be using TCP keepalive to force some traffic over the connection
+ * periodically. */
+#define PROXY_DEFAULT_LIFETIME 86400
```
Comment (rlenglet) [inline bpf/lib/common.h:404]: Does this lifetime need to be that long? With https://github.com/cilium/cilium/pull/2860, we'll have a guarantee that Kafka connections don't live longer than ~1 minute after a redirect is removed. I don't understand yet the behavior of Envoy, but we can probably set a shorter lifetime.

**Review by rlenglet — APPROVED** [review]

File: pkg/proxy/dialer.go (hunk 1 above)
Comment (rlenglet) [reply pkg/proxy/dialer.go]: I'll set the SO_LINGER option on the request connection in https://github.com/cilium/cilium/pull/2860. I think that's required at least on request side.

**Comment by rlenglet** [comment]
FYI, the Kafka broker is already setting the keepalive option on all accepted sockets: https://github.com/apache/kafka/commit/194d1fc030561fd06e39224584b58cc49e2f63e8.

File: bpf/lib/common.h (hunk 2 above)
Comment (tgraf) [reply bpf/lib/common.h:404]: I'm happy to adjust this down again once we understand the behaviour of Envoy.

File: pkg/proxy/dialer.go (hunk 1 above)
Comment (tgraf) [reply pkg/proxy/dialer.go]: fixed

**Comment by tgraf** [comment]
test-me-please

**Comment by tgraf** [comment]
test-me-please

**Comment by tgraf** [comment]
test-me-please

**Comment by tgraf** [comment]
test-me-please

**Comment by tgraf** [comment]
test-me-please

**Comment by tgraf** [comment]
test-me-please

**Comment by tgraf** [comment]
test-me-please

**Comment by tgraf** [comment]
test-me-please

**Pull request merged**